packages/api/src/middleware/validation/checkAnimalOrBatch.js

```diff
@@ -13,81 +13,606 @@
+import { oneExists, setFalsyValuesToNull } from '../../util/middleware.js';
+import {
+  customError,
+  checkIsArray,
+  checkIdIsNumber,
+  checkExactlyOneIsProvided,
+  checkRecordBelongsToFarm,
+} from '../../util/customErrors.js';
+
 import AnimalModel from '../../models/animalModel.js';
 import AnimalBatchModel from '../../models/animalBatchModel.js';
+import CustomAnimalTypeModel from '../../models/customAnimalTypeModel.js';
+import DefaultAnimalBreedModel from '../../models/defaultAnimalBreedModel.js';
+import CustomAnimalBreedModel from '../../models/customAnimalBreedModel.js';
+import AnimalUseModel from '../../models/animalUseModel.js';
+import AnimalOriginModel from '../../models/animalOriginModel.js';
+import AnimalIdentifierType from '../../models/animalIdentifierTypeModel.js';
 
 const AnimalOrBatchModel = {
   animal: AnimalModel,
   batch: AnimalBatchModel,
 };
 
-export function checkEditAnimalOrBatch(animalOrBatchKey) {
+const checkValidAnimalOrBatchIds = async (animalOrBatchKey, ids, farm_id, trx) => {
+  if (!ids || !ids.length) {
+    throw customError('Must send ids');
+  }
+
+  const idsSet = new Set(ids.split(','));
+
+  // Check that all animals/batches exist and belong to the farm
+  const invalidIds = [];
+
+  for (const id of idsSet) {
+    // For query syntax like ids=,,, which will pass the above check
+    checkIdIsNumber(id);
+
+    const existingRecord = await AnimalOrBatchModel[animalOrBatchKey]
+      .query(trx)
+      .findById(id)
+      .where({ farm_id })
+      .whereNotDeleted(); // prohibiting re-delete
+
+    if (!existingRecord) {
+      invalidIds.push(id);
+    }
+  }
+
+  if (invalidIds.length) {
+    throw customError(
+      'Some entities do not exist, are already deleted, or are not associated with the given farm.',
+      400,
+      { error: 'Invalid ids', invalidIds },
+    );
+  }
+};
+
+// For edit mode set creating to false
+const checkAnimalType = async (animalOrBatch, farm_id, creating = true) => {
+  const { default_type_id, custom_type_id, type_name } = animalOrBatch;
+  const typeKeyOptions = ['default_type_id', 'custom_type_id', 'type_name'];
+  // Skip if all undefined or editing (!creating)
+  if (creating || oneExists(typeKeyOptions, animalOrBatch)) {
+    checkExactlyOneIsProvided(
+      [default_type_id, custom_type_id, type_name],
+      'default_type_id, custom_type_id, or type_name',
+    );
+    if (!creating) {
+      // Overwrite with null in db if editing
+      setFalsyValuesToNull(typeKeyOptions, animalOrBatch);
+    }
+    if (custom_type_id) {
+      checkIdIsNumber(custom_type_id);
+      const customType = await CustomAnimalTypeModel.query().findById(custom_type_id);
+      if (!customType) {
+        throw customError('Custom type does not exist');
+      }
+      await checkRecordBelongsToFarm(customType, farm_id, 'custom type');
+    }
+  }
+};
+
+const checkDefaultBreedMatchesType = async (
+  animalOrBatchRecord,
+  default_breed_id,
+  default_type_id,
```

Comment:
If I passed animalOrBatch and used it directly in the next lines would that be better? I kept the name to represent the destructured from animalOrBatch.

Reply:
If you passed `animalOrBatch`, I think that would make it clearer. I wasn't sure where those came from since I reviewed this function first.
But I'm used to it now, and it would be a lot of work, so no need to change!